File: Source/Binary/Creators.Collections/ListConverter.cs
using Mikodev.Binary.Abstractions;
using Mikodev.Binary.Adapters;
using System;
using System.Collections.Generic;

namespace Mikodev.Binary.Creators.Collections
{
    internal sealed class ListConverter<T> : VariableConverter<List<T>>
    {
        private readonly Adapter<T> adapter;

        public ListConverter(Converter<T> converter) => adapter = AdapterHelper.Create(converter);

        public override void ToBytes(ref Allocator allocator, List<T> item) => adapter.OfList(ref allocator, item);

        public override List<T> ToValue(in ReadOnlySpan<byte> span) => adapter.ToList(in span);
    }
}

File: Source/Binary.FSharp/Internal.Creators.Collections/ListConverter.fs
namespace Mikodev.Binary.Internal.Creators.Collections

open Mikodev.Binary
open System

[<CompiledName("FSharpListConverter`1")>]
type ListConverter<'T>(converter : Converter<'T>, memoryConverter : Converter<Memory<'T>>) =
    inherit Converter<List<'T>>(0)

    override __.Encode(allocator, item) =
        if isNull (box item) = false then
            let converter = converter
            for i in item do
                converter.EncodeAuto(&allocator, i)
        ()

    override __.Decode(span : inref<ReadOnlySpan<byte>>) : List<'T> =
        let data = (memoryConverter.Decode &span).Span
        let mutable list = []
        for i = data.Length - 1 downto 0 do
            list <- data.[i] :: list
        list
